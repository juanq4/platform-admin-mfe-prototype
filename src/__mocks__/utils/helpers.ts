import { isEmpty } from "@q4/nimbus-ui";

export function mockWindowMethods(): Window {
  return Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
}

export function mockClipboardWrite(): void {
  const orginalClipboard = { ...navigator.clipboard };

  beforeAll(() => {
    if (!isEmpty(navigator.clipboard)) return;

    Object.defineProperty(navigator, "clipboard", {
      value: {
        writeText: () => Promise.resolve(),
      },
    });
  });

  afterAll(() => {
    Object.assign(navigator.clipboard, {
      ...orginalClipboard,
    });
  });
}
