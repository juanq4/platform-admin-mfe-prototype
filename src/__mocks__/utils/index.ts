export * from "./helpers";
export * from "./permissions";
export * from "./wrappers";
