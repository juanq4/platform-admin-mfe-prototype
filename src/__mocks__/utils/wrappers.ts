export const getAppWrapper = (): HTMLElement => {
  return global.document.getElementById("root");
};
