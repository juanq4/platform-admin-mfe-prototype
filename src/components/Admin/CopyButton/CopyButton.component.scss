@import "/src/styles/scss/variables";
@import "/src/styles/scss/mixins";

.copy-button.nui-button--v1 {
  padding: gux(1.5);
}
