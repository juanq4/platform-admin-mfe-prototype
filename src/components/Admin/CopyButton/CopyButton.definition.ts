import { ButtonIdModel } from "@q4/nimbus-ui";
import type { BaseComponentProps } from "@q4/nimbus-ui";

export interface CopyButtonProps extends Pick<BaseComponentProps, "id" | "key"> {
  value: string;
  label: string;
}

export enum CopyButtonClassName {
  Base = "copy-button",
}

export const CopyButtonDefault = {
  Label: "value",
};

export { ButtonIdModel as CopyButtonIdModel };
