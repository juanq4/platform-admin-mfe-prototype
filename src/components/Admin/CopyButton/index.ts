export * from "./CopyButton.component";
export * from "./CopyButton.definition";
export * from "./CopyButton.utils";
