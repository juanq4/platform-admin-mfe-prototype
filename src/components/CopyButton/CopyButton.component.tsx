import "./CopyButton.scss";
import { Button, ButtonTheme, NotificationService } from "@q4/nimbus-ui";
import { memo, useCallback, useMemo, useRef } from "react";
import type { CopyButtonProps } from "./CopyButton.definition";
import { CopyButtonClassName, CopyButtonIdModel } from "./CopyButton.definition";
import { getCopyButtonFailureMessage, getCopyButtonSuccessMessage, isCopyButtonDisabled } from "./CopyButton.utils";

const CopyButtonBase = (props: CopyButtonProps): JSX.Element => {
  const notificationServiceRef = useRef(new NotificationService());

  const { id, label, value } = props;

  const idModel = useMemo(() => new CopyButtonIdModel(id), [id]);
  const disabled = useMemo(() => isCopyButtonDisabled(value), [value]);

  const copyToClipboard = useCallback((): void => {
    const { current: notificationService } = notificationServiceRef;

    notificationService.dismiss();
    navigator.clipboard
      .writeText(value)
      .then(() => {
        notificationService.success(getCopyButtonSuccessMessage(label));
      })
      .catch(() => {
        notificationService.error(getCopyButtonFailureMessage(label));
      });
  }, [label, value]);

  return (
    <Button
      id={idModel.id}
      className={CopyButtonClassName.Base}
      theme={ButtonTheme.None}
      icon="q4i-clone-duplicate-4pt"
      disabled={disabled}
      stopPropagation
      onClick={copyToClipboard}
    />
  );
};
export const CopyButton = memo(CopyButtonBase);
