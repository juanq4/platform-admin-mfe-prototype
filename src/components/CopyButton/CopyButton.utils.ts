import { isNullOrWhiteSpace } from "@q4/nimbus-ui";
import { CopyButtonDefault } from "./CopyButton.definition";
import type { CopyButtonProps } from "./CopyButton.definition";

export function getCopyButtonSuccessMessage(label: CopyButtonProps["label"]): string {
  const labelValue = _getDefaultLabel(label);
  return `The ${labelValue} has been copied to your clipboard.`;
}

export function getCopyButtonFailureMessage(label: CopyButtonProps["label"]): string {
  const labelValue = _getDefaultLabel(label);
  return `The ${labelValue} failed to copy to your clipboard.`;
}

export function isCopyButtonDisabled(value: CopyButtonProps["value"]): boolean {
  return isNullOrWhiteSpace(value);
}

function _getDefaultLabel(label: CopyButtonProps["label"]) {
  return isNullOrWhiteSpace(label) ? CopyButtonDefault.Label : label;
}
