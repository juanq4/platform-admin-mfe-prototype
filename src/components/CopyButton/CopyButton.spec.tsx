import { ToastContainer } from "@q4/nimbus-ui";
import React from "react";
import { NimbusConfig } from "../../__mocks__/contexts/NimbusConfig.mock";
import { mockClipboardWrite } from "../../__mocks__/utils/helpers";
import { getAppWrapper } from "../../__mocks__/utils/wrappers";
import { fireEvent, render, screen, waitFor } from "../../utils/testUtils";
import { CopyButton } from "./CopyButton.component";
import type { CopyButtonProps } from "./CopyButton.definition";
import { CopyButtonDefault, CopyButtonIdModel } from "./CopyButton.definition";
import { getCopyButtonFailureMessage, getCopyButtonSuccessMessage, isCopyButtonDisabled } from "./CopyButton.utils";

const MockLabel = "mock label";
const MockValue = "this is a mock string";

describe("CopyButton Component", () => {
  mockClipboardWrite();

  function renderWithConfigs(props: CopyButtonProps) {
    return (
      <NimbusConfig.ConfigProvider>
        <CopyButton {...props} />
      </NimbusConfig.ConfigProvider>
    );
  }

  test("7378438: [Given] no props are provided [Expect] the component to render", () => {
    const props: CopyButtonProps = {
      value: null,
      label: null,
    };
    render(renderWithConfigs(props), { container: getAppWrapper() });
    const rootElement = screen.queryByTestId("root");
    expect(rootElement).toMatchSnapshot();
  });

  const idModel = new CopyButtonIdModel("MockCopyButtonId");
  const { id } = idModel;
  const mockProps = [
    {
      id,
      value: MockValue,
      label: MockLabel,
      expectedLabel: MockLabel,
    },
    {
      id,
      value: MockValue,
      label: null as string,
      expectedLabel: CopyButtonDefault.Label,
    },
  ];

  test("7378439: [Given] the copy the text value is 'null' [Expect] the button to be disabled", () => {
    const props = {
      id,
      value: null as string,
      label: MockLabel,
    };

    render(renderWithConfigs(props));

    const button = screen.queryByTestId(id);
    expect(button).toBeDisabled();
    expect(button).toMatchSnapshot();
  });

  test("7449116: [Given] copying the text 'this is a mock string' was a success [And] the label is 'mock label' [Expect] a success toast message with 'mock label'  to appear", async () => {
    const { expectedLabel, ...componentProps } = mockProps[0];
    Object.assign(navigator.clipboard, { writeText: () => Promise.resolve() });

    render(
      <>
        {renderWithConfigs(componentProps)}
        <ToastContainer />
      </>,
    );

    const button = screen.queryByTestId(id);
    expect(button).toBeInTheDocument();
    fireEvent.click(button);

    // FIXME: PLATFORM-1926 This type of assertion is very slow. Instead, the test should
    // be mocking useNotifications and putting an assertion on notifications.current.success.
    // Once that the invocation of the notification is asserted, there would be no need
    // for a snapshot test either.
    let toast: Element;

    await waitFor(() => {
      toast = screen.getByText(getCopyButtonSuccessMessage(expectedLabel))?.parentElement?.parentElement;
      expect(toast).toBeInTheDocument();
    });
  });

  test("7449117: [Given] copying the text 'this is a mock string' was a success [And] the label is 'null' [Expect] a success toast message with 'value' to appear", async () => {
    const { expectedLabel, ...componentProps } = mockProps[1];
    Object.assign(navigator.clipboard, { writeText: () => Promise.resolve() });

    render(
      <>
        {renderWithConfigs(componentProps)}
        <ToastContainer />
      </>,
    );

    const button = screen.queryByTestId(id);
    expect(button).toBeInTheDocument();
    fireEvent.click(button);

    // FIXME: PLATFORM-1926 This type of assertion is very slow. Instead, the test should
    // be mocking useNotifications and putting an assertion on notifications.current.success.
    // Once that the invocation of the notification is asserted, there would be no need
    // for a snapshot test either.
    let toast: Element;

    await waitFor(() => {
      toast = screen.getByText(getCopyButtonSuccessMessage(expectedLabel))?.parentElement?.parentElement;
      expect(toast).toBeInTheDocument();
    });
  });

  test("7449118: [Given] copying the text 'this is a mock string' was a failure [And] the label is 'mock label' [Expect] a failure toast message with 'mock label'  to appear", async () => {
    const { expectedLabel, ...componentProps } = mockProps[0];
    Object.assign(navigator.clipboard, { writeText: () => Promise.reject() });

    render(
      <>
        {renderWithConfigs(componentProps)}
        <ToastContainer />
      </>,
    );

    const button = screen.queryByTestId(id);
    expect(button).toBeInTheDocument();
    fireEvent.click(button);

    // FIXME: PLATFORM-1926 This type of assertion is very slow. Instead, the test should
    // be mocking useNotifications and putting an assertion on notifications.current.success.
    // Once that the invocation of the notification is asserted, there would be no need
    // for a snapshot test either.
    let toast: Element;

    await waitFor(() => {
      toast = screen.getByText(getCopyButtonFailureMessage(expectedLabel))?.parentElement?.parentElement;
      expect(toast).toBeInTheDocument();
    });
  });

  test("7449119: [Given] copying the text 'this is a mock string' was a failure [And] the label is 'mock label' [Expect] a failure toast message with 'mock label'  to appear", async () => {
    const { expectedLabel, ...componentProps } = mockProps[1];
    Object.assign(navigator.clipboard, { writeText: () => Promise.reject() });

    render(
      <>
        {renderWithConfigs(componentProps)}
        <ToastContainer />
      </>,
    );

    const button = screen.queryByTestId(id);
    expect(button).toBeInTheDocument();
    fireEvent.click(button);

    // FIXME: PLATFORM-1926 This type of assertion is very slow. Instead, the test should
    // be mocking useNotifications and putting an assertion on notifications.current.success.
    // Once that the invocation of the notification is asserted, there would be no need
    // for a snapshot test either.
    let toast: Element;

    await waitFor(() => {
      toast = screen.getByText(getCopyButtonFailureMessage(expectedLabel))?.parentElement?.parentElement;
      expect(toast).toBeInTheDocument();
    });
  });
});

test(`7449121: [Given] the label is '${MockLabel}' [Expect] the returned label to match'`, () => {
  expect(getCopyButtonSuccessMessage(MockLabel)).toBe("The mock label has been copied to your clipboard.");
});

test("7449122: [Given] the label is '' [Expect] the returned label to match'", () => {
  expect(getCopyButtonSuccessMessage("")).toBe("The value has been copied to your clipboard.");
});

test(`7449691: [Given] the label is '${MockLabel}' [Expect] the returned label to match'`, () => {
  expect(getCopyButtonFailureMessage(MockLabel)).toBe("The mock label failed to copy to your clipboard.");
});

test("7449692: [Given] the label is '' [Expect] the returned label to match'", () => {
  expect(getCopyButtonFailureMessage("")).toBe("The value failed to copy to your clipboard.");
});

test("7449693: [Given] the label is 'this is a mock string' [Expect] the returned label to match'", () => {
  const returned = isCopyButtonDisabled(MockValue);
  expect(returned).toMatchSnapshot();
});

test("7449694: [Given] the label is 'null' [Expect] the returned label to match'", () => {
  const returned = isCopyButtonDisabled(null);
  expect(returned).toMatchSnapshot();
});

describe("isCopyButtonDisabled", () => {
  test("7449693: [Given] the label is 'this is a mock string' [Expect] the returned label to match'", () => {
    const returned = isCopyButtonDisabled(MockValue);
    expect(returned).toMatchSnapshot();
  });

  test("7449694: [Given] the label is 'null' [Expect] the returned label to match'", () => {
    const returned = isCopyButtonDisabled(null);
    expect(returned).toMatchSnapshot();
  });
});
