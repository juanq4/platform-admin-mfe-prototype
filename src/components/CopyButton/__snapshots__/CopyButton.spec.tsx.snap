// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`7449693: [Given] the label is 'this is a mock string' [Expect] the returned label to match' 1`] = `false`;

exports[`7449694: [Given] the label is 'null' [Expect] the returned label to match' 1`] = `true`;

exports[`CopyButton Component 7378438: [Given] no props are provided [Expect] the component to render 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  background-color: #0f5ca3;
  color: #ffffff;
  border-color: transparent;
  font-family: inherit;
  height: 40px;
  padding: 12px 16px;
  border-width: 1px;
  border-style: solid;
  font-size: 14px;
  line-height: 16px;
  text-transform: none;
  -webkit-transition: 0.3s ease-in-out;
  transition: 0.3s ease-in-out;
  cursor: pointer!important;
  border-radius: 4px;
  padding: 0;
  width: 40px;
  text-align: center;
}

.emotion-0:hover {
  background: #0a3d6d;
  color: #ffffff;
}

.emotion-0:hover [data-type="nui-tooltip"] {
  opacity: 1;
  visibility: visible;
}

.emotion-0:focus {
  background: #0a3d6d;
  border-color: #4ea1ef;
  box-shadow: 0px 0px 3px #4ea1ef;
}

.emotion-0:active {
  background: #051f36;
}

.emotion-0:disabled {
  pointer-events: none;
  cursor: default;
  background: #f7fafc;
  color: #61707c;
  border-color: transparent;
}

@media only screen and (max-width: 768px) {
  .emotion-0 {
    height: 40px;
    padding: 12px;
    width: 40px;
  }
}

.emotion-1 {
  font-size: 12px;
  line-height: 40px;
}

.emotion-1+span {
  margin-left: 4px;
}

@media only screen and (max-width: 768px) {
  .emotion-1 {
    line-height: 16px;
  }
}

<div
  id="root"
>
  <button
    class="copy-button emotion-0"
    disabled=""
    type="button"
  >
    <i
      class="copy-button_icon q4i-clone-duplicate-4pt emotion-1"
    />
  </button>
</div>
`;

exports[`CopyButton Component 7378439: [Given] the copy the text value is 'null' [Expect] the button to be disabled 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  background-color: #0f5ca3;
  color: #ffffff;
  border-color: transparent;
  font-family: inherit;
  height: 40px;
  padding: 12px 16px;
  border-width: 1px;
  border-style: solid;
  font-size: 14px;
  line-height: 16px;
  text-transform: none;
  -webkit-transition: 0.3s ease-in-out;
  transition: 0.3s ease-in-out;
  cursor: pointer!important;
  border-radius: 4px;
  padding: 0;
  width: 40px;
  text-align: center;
}

.emotion-0:hover {
  background: #0a3d6d;
  color: #ffffff;
}

.emotion-0:hover [data-type="nui-tooltip"] {
  opacity: 1;
  visibility: visible;
}

.emotion-0:focus {
  background: #0a3d6d;
  border-color: #4ea1ef;
  box-shadow: 0px 0px 3px #4ea1ef;
}

.emotion-0:active {
  background: #051f36;
}

.emotion-0:disabled {
  pointer-events: none;
  cursor: default;
  background: #f7fafc;
  color: #61707c;
  border-color: transparent;
}

@media only screen and (max-width: 768px) {
  .emotion-0 {
    height: 40px;
    padding: 12px;
    width: 40px;
  }
}

.emotion-1 {
  font-size: 12px;
  line-height: 40px;
}

.emotion-1+span {
  margin-left: 4px;
}

@media only screen and (max-width: 768px) {
  .emotion-1 {
    line-height: 16px;
  }
}

<button
  class="copy-button emotion-0"
  disabled=""
  id="MockCopyButtonId"
  type="button"
>
  <i
    class="copy-button_icon q4i-clone-duplicate-4pt emotion-1"
    id="MockCopyButtonIdIcon"
  />
</button>
`;

exports[`isCopyButtonDisabled 7449693: [Given] the label is 'this is a mock string' [Expect] the returned label to match' 1`] = `false`;

exports[`isCopyButtonDisabled 7449694: [Given] the label is 'null' [Expect] the returned label to match' 1`] = `true`;
